rust/appencryption/examples/custom_metastore.rs

```diff
@@ -0,0 +1,245 @@
+use appencryption::{
+    envelope::EnvelopeKeyRecord,
+    kms::StaticKeyManagementService,
+    policy::CryptoPolicy,
+    session::{Session, SessionFactory},
+    Metastore,
+};
+use async_trait::async_trait;
+use securememory::protected_memory::DefaultSecretFactory;
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+use std::time::{SystemTime, UNIX_EPOCH};
+
+/// Example demonstrating how to implement a custom Metastore for Asherah.
+///
+/// This example shows:
+/// 1. How to create a custom Metastore implementation
+/// 2. How to integrate the custom Metastore with Asherah
+/// 3. Basic encrypt/decrypt operations using the custom Metastore
+
+/// A simple Redis-like Metastore implementation.
+///
+/// This is a simplified example that stores keys in memory with Redis-like TTL features.
+/// In a real application, this would connect to a Redis server.
+struct RedisMetastore {
+    /// In-memory storage map for keys
+    store: RwLock<HashMap<String, (EnvelopeKeyRecord, i64, Option<i64>)>>,
+}
+
+impl RedisMetastore {
+    /// Create a new RedisMetastore
+    fn new() -> Self {
+        Self {
+            store: RwLock::new(HashMap::new()),
+        }
+    }
+
+    /// Generate a key string from id and created timestamp
+    fn generate_key(id: &str, created: i64) -> String {
+        format!("{}_{}", id, created)
+    }
+
+    /// Set a TTL (time-to-live) for a key in seconds
+    async fn expire(&self, id: &str, created: i64, ttl_seconds: i64) -> bool {
```

Comment:
I don't know enough about the inner workings of asherah, but it seems having a generic Trait that predefines the requires get/set/delete/expire methods for key/value store backings would be useful. It would enforce every metastore definition has async and has proper error handling where needed.

Generic crates that handle these abstractions also exist, but I don't know what the quality of those is.